### PR TITLE
.Net Concurrency

### PR DESCRIPTION
**Pull request by nesalang**
## Related Issue
N/A
## Description
Opening this PR to continue the CR.
Please take a note that this is a topical branch from jwoo/dot-net-concur which has been merged.
CI tests are not expected to pass; once CR is done, the base branch will be updated to master.

Thanks

ParseContext was a static class, and it caused synchronization issue during parsing in multi-threaded environments.
this changes include:
Made ParseContext non static class
Added an instance of ParseContext to WarningLoggingContractResolver.
so client's custom Json contract resolver may not work.
we already don't allow custom Json contract resolver if FromJson() API is used
Fixed the issue where internalIds are not being set for elements
Added concurrency unit test for parsing

## How Verified
Concurrency unit tests were added and verified it.
All existing unit tests were passed
Run all of the fallback json tests cards and ran some of the relevant cards such as showcard

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/3556)

### CODE REVIEW

File: source/dotnet/Test/AdaptiveCards.Test/AdaptiveCardApiTests.cs [hunk 1]
```diff
@@ -1245,5 +1245,156 @@ public void TestParsingRichTextBlockWithInvalidInlineType()
...
+
+            try
+            {
+                Task.WaitAll(tasks.ToArray());
+            }
+            catch
+            {
+                // it's perfectly fine card. if there is exception, it is due to concurreny
+                // as it's the only variable.
+                Assert.Fail("test tried to parse a fine card in multi threads, and failed, please check if thread safety is broken");
```
Comment (shalinijoshi19) [inline source/dotnet/Test/AdaptiveCards.Test/AdaptiveCardApiTests.cs]: Nit: "Unexpected failure parsing a valid AdaptiveCard" should suffice

**Comment by shalinijoshi19** [comment]
>                 // internal id in dot net needs to be revisited tracked via issue 3386

At some point we should be revisiting some of these TODO issues mentioned in the code base (note to self mostly :) ) 

---
Refers to: source/dotnet/Library/AdaptiveCards/ParseContext.cs:32 in 49f9297. [](commit_id = 49f92977d4bbb79c2f5ca6c2262bfbe3740d4162, deletion_comment = False)

**Comment by shalinijoshi19** [comment]
>             AdaptiveFallbackElement fallback = value as AdaptiveFallbackElement;

This file is missing a copyright header; Could you add that in.. thanks! 

---
Refers to: source/dotnet/Library/AdaptiveCards/AdaptiveFallbackConverter.cs:15 in 49f9297. [](commit_id = 49f92977d4bbb79c2f5ca6c2262bfbe3740d4162, deletion_comment = False)

File: source/dotnet/Library/AdaptiveCards/ParseContext.cs [hunk 2]
```diff
@@ -5,24 +5,18 @@
...
-            idStack = new Stack<Tuple<string, AdaptiveInternalID, bool>>();
-        }
-
         public enum ContextType { Element, Action };
 
-        public static ContextType Type { get; set; }
+        public ContextType Type { get; set; }
 
-        private static IDictionary<string, List<AdaptiveInternalID>> elementIds = new Dictionary<string, List<AdaptiveInternalID>>();
+        private IDictionary<string, List<AdaptiveInternalID>> elementIds = new Dictionary<string, List<AdaptiveInternalID>>();
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards/ParseContext.cs]: This file is also missing hte copyright header. Could you add it in? thanks!

Comment (nesalang) [reply source/dotnet/Library/AdaptiveCards/ParseContext.cs]: sure

---
In reply to: [341394563](https://github.com/microsoft/AdaptiveCards/pull/3556#discussion_r341394563) [](ancestors = 341394563)

**Comment by nesalang** [comment]
>                 // internal id in dot net needs to be revisited tracked via issue 3386

o.k.

---
In reply to: [548592546](https://github.com/microsoft/AdaptiveCards/pull/3556#issuecomment-548592546) [](ancestors = 548592546)

---
Refers to: source/dotnet/Library/AdaptiveCards/ParseContext.cs:32 in 49f9297. [](commit_id = 49f92977d4bbb79c2f5ca6c2262bfbe3740d4162, deletion_comment = False)

**Comment by nesalang** [comment]
>             AdaptiveFallbackElement fallback = value as AdaptiveFallbackElement;

sure

---
In reply to: [548597068](https://github.com/microsoft/AdaptiveCards/pull/3556#issuecomment-548597068) [](ancestors = 548597068)

---
Refers to: source/dotnet/Library/AdaptiveCards/AdaptiveFallbackConverter.cs:15 in 49f9297. [](commit_id = 49f92977d4bbb79c2f5ca6c2262bfbe3740d4162, deletion_comment = False)

File: source/dotnet/Library/AdaptiveCards/AdaptiveInlinesConverter.cs [hunk 3]
```diff
@@ -24,6 +24,11 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
             var array = JArray.Load(reader);
             List<object> list = array.ToObject<List<object>>();
             List<AdaptiveInline> arrayList = new List<AdaptiveInline>();
+            var serializerSettigns = new JsonSerializerSettings
+            {
+                ContractResolver = new WarningLoggingContractResolver(new AdaptiveCardParseResult(), ParseContext),
+                Converters = { new StrictIntConverter() }
```
Comment (shalinijoshi19) [inline source/dotnet/Library/AdaptiveCards/AdaptiveInlinesConverter.cs]: >StrictIntConverter [](start = 35, length = 18)

Question: Why this?

Comment (nesalang) [reply source/dotnet/Library/AdaptiveCards/AdaptiveInlinesConverter.cs]: I don't have particular reason for using the converter. I'm following the usage shown in AdaptiveCard.cs LN239-242. I couldn't find reasons to differ.

---
In reply to: [341414800](https://github.com/microsoft/AdaptiveCards/pull/3556#discussion_r341414800) [](ancestors = 341414800)

**Pull request merged**